text/0000-build-std.md

```diff
@@ -0,0 +1,4582 @@
+> with `-L dependency=` and then from the sysroot. rustc will attempt to load
+> some crates like `compiler_builtins` and `panic_unwind` itself.
+>
+> rustc will need patched to be able to load `panic_unwind` from
+> `-L dependency=` paths.
+>
+> Cargo will load the sysroot crate in the standard library and perform the
+> resolve on that workspace with the packages that the user has a explicit
+> dependency on and those which require `--extern` to be passed explicitly.
+
+The host pre-built standard library will always be used for procedural macros
+and build scripts ([?][rationale-sysroot-for-host-deps]). Artifact dependencies
+use the same standard library as the rest of the crate (pre-built or
+newly-built, as appropriate).
+
+*See the following sections for rationale/alternatives:*
+
+- [*Why put `build-std` in the Cargo config?*][rationale-build-std-in-config]
+- [*Why not always rebuild when the profile changes?*][rationale-why-not-always-rebuild]
+- [*Why have different build-std defaults depending on the profile?*][rationale-different-defaults]
+- [*Why rebuild the standard library automatically?*][rationale-why-automatic]
+- [*Why use the lockfile of the `rust-src` component?*][rationale-lockfile]
+- [*Why not build the standard library in incremental?*][rationale-incremental]
+- [*Why not produce a `dylib` for the standard library?*][rationale-no-dylib]
+- [*Why use the pre-built standard library for procedural macros and build-scripts?*][rationale-sysroot-for-host-deps]
+
+*See the following sections for relevant unresolved questions:*
+
+- [*Where should the `build-std` configuration in `.cargo/config` be and what should it be called?*][unresolved-config-location-name]
+- [*What should the values of the `build-std` config be named?*][unresolved-config-values]
+
+### Profiles
+[profiles]: #profiles
+
+Cargo will assume that the pre-built standard library matches the standard
+library's release profile ([?][rationale-assume-release-profile]). If the user
+changes the default release profile or builds with a different profile then this
+could trigger a rebuild of the standard library
+([?][rationale-ship-debug-std]), depending on the value of the `build-std`
+config as above.
+
+User's Cargo profile | Target modifiers changed? | Standard library profile
+-------------------- | ------------------------- | ------------------------
+`release`            | No                        | `release` (pre-built)
+`dev`                | No                        | `release` (pre-built)
+`release`            | Yes                       | `release` (newly built)
+`dev`                | Yes                       | `dev` (newly built)
+
+> [!NOTE]
+>
+> If `build-std` is set to `target-modifiers`, Cargo must decide if `build-std`
+> should be enabled. rustc will add a `--print target-modifiers` flag which will
+> print all of the flags treated as target modifiers, like `-Zretpoline`, with
```

Comment:
Cargo has no knowledge of these flags, so I'm not sure how this would work.

Reply:
I imagined that Cargo would have a list of the flags that it was passing to any given rustc invocation, and it could - without any knowledge of what the flags do - compare them to a list of target modifiers flags provided by rustc, and if the list overlaps then it needs a rebuild.

I'm not the biggest fan of that mechanism though, admittedly. @wesleywiser and I discussed adding some compiler support here, so that Cargo could invoke rustc with the rustc flags derived from the user's profile and with the path to the sysroot's `libstd.rlib` and rustc would tell you if the target modifiers match. As the profile is the same for the entire workspace, Cargo would do this once for the current profile.

Reply:
I like the probe approach as it gives RUSTFLAGS support for free as there's precedent for other build probes [inheriting RUSTFLAGS](https://github.com/rust-lang/cargo/blob/e4162389d67c603d25ba6e25b0e9423fcb8daa64/src/cargo/core/compiler/build_context/target_info.rs#L770), at least when in cross-compile mode.

The probe could happen after the profile is constructed as it looks like they're built before the resolver runs.

It's possible for profile settings to differ for different crates, but if these differ by a target modifier then that will break anyway.